tests/FakeItEasy.Specs/ArgumentCapturingSpecs.cs

```diff
@@ -0,0 +1,318 @@
+namespace FakeItEasy.Specs
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using FakeItEasy.Tests.TestHelpers;
+    using FluentAssertions;
+    using Xbehave;
+    using Xunit;
+
+    public static class ArgumentCapturingSpecs
+    {
+        [Scenario]
+        public static void CaptureSingleValue(Action<int> fake, Captured<int> capturedArgument)
+        {
+            "Given a fake"
+                .x(() => fake = A.Fake<Action<int>>());
+
+            "And a captured argument instance"
+                .x(() => capturedArgument = A.Captured<int>());
+
+            "And a fake method is configured to capture its argument to that instance"
+                .x(() => A.CallTo(() => fake.Invoke(capturedArgument._))
+                            .DoesNothing());
+
+            "When the method is called"
+                .x(() => fake.Invoke(354897));
+
+            "Then the captured argument instance's last value matches the input"
+                .x(() => capturedArgument.GetLastValue().Should().Be(354897));
+        }
+
+        [Scenario]
+        public static void MismatchedVerifyWithCapture(
+            Action<int, int> fake, Captured<int> capturedArgument, Exception exception)
+        {
+            "Given a fake"
+                .x(() => fake = A.Fake<Action<int, int>>());
+
+            "And a captured argument instance"
+                .x(() => capturedArgument = A.Captured<int>());
+
+            "And a method on the fake is called"
+                .x(() => fake.Invoke(354897, -3));
+
+            "When we assert that the call must've happened with a capturing argument and another non-matching argument"
+                .x(() => exception = Record.Exception(
+                    () => A.CallTo(() => fake.Invoke(capturedArgument._, 3))
+                            .MustHaveHappened()));

```

Comment:
This is actually a use case that I didn't consider before. Does it even make sense to use a `Captured` in a call assertion? It won't capture the value, since the call has already happened...
Not sure if we could/should do something to prevent it.

Reply:
I don't think _I'd_ ever do it. I guess the more likely scenario is that someone saves away the result of `A.CallTo`. I do do that. I can rewrite the test to use that flow.

Is it likely to capture and use `MustHaveHappened`? I dunno. But if people do it (and I don't think we should prevent it), it'd be nice if it works. 😄

Reply:
> I can rewrite the test to use that flow.

No, don't bother.

> Is it likely to capture and use `MustHaveHappened`? I dunno. But if people do it (and I don't think we should prevent it), it'd be nice if it works. 😄

What would "works" entail? It can't capture the value, since the call already happened.

Reply:
Oh, capture, but save away the result of `A.CallTo` to use later with a `MustHavHappened`. I'd expect any constraints applied to the `Captured` to be used correctly in the `MustHaveHappened`.

```c#
var capturedArg = A.Captured<int>();

var aCallToMethod = A.CallTo(() => fake.Method(capturedArg.That.IsDivisibleBy7());

aCallToMethod.Returns(6);

…

// passes if there was one call that supplied an argument divisible by 7
aCallToMethod.MustHaveHappened(); 
```

Reply:
Ah, right. Sounds good to me.